### PR TITLE
add PIDs stats to exported prometheus container metrics

### PR DESCRIPTION
**Pull request by adrien-f**
Greetings !

We were in need of tracking global PIDs usage across our containers via Prometheus so we set out forking and adding this little feature. We're running this new feature in production (easy debian repackaging !) and we're really satisfied so far. 

cadvisor-container-stats

As this is my first PR for this project, let me know if you need more informations or tests.

Thanks :wave: 

Signed-off-by: Adrien Fillon <adrien.fillon@cdiscount.com>

### CODE REVIEW

**Comment by adrien-f** [comment]
Thanks for the feedback, I'll look into it and get back to you !

**Comment by adrien-f** [comment]
Alright, I've update the pull request, `pids` should appear in all versions of the API, is that alright ?

I've been a bit troubled with the recent changes with `container/common` and `container/libcontainer`, let me know if I forgot anything.

**Comment by adrien-f** [comment]
Is there something wrong with the e2e tests or is it my PR ?

**Comment by dashpole** [comment]
It looks like the tests are a little flaky right now: https://k8s-testgrid.appspot.com/sig-node-cadvisor#cadvisor-e2e, but since it is failing consistently, it is probably something in your PR

**Comment by dashpole** [comment]
Looks like all tests are failing with this on each distro: 
`I0228 08:56:02.077] --- FAIL: TestMachineStatsIsReturned (0.01s)`
`I0228 08:56:02.077] 	machinestats_test.go:32: unable to post "machine stats" to "http://localhost:8080/api/v2.1/machinestats": Get http://localhost:8080/api/v2.1/machinestats: EOF`

There is a panic in one of the cAdvisor logs: https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/google_cadvisor/1891/pull-cadvisor-e2e/689/artifacts/tmp-node-e2e-db502bef-cos-stable-60-9592-84-0/cadvisor.log

**Comment by adrien-f** [comment]
I've nailed down the issue to cadvisor requesting stats for the root container `/` that resolves to `/sys/fs/cgroup/`. Trying to get its pids info in `/vendor/github.com/opencontainers/runc/libcontainer/cgroups/fs/pids.go` leads to an error since the `pids.current` file does not exist.

```
Failed to update stats for container "/": failed to parse pids.current - open /sys/fs/cgroup/pids/pids.current: no such file or directory, continuing to push stats
```
I'm still looking on how to avoid this behaviour.

**Comment by adrien-f** [comment]
I'm not sure I took the best approach here, let me know what you think :+1:

**Comment by dashpole** [comment]
Can you explain the issue a bit more?  Is it that certain OS distros don't have the file, or that the root "/" cgroup doesn't have this file?  Or something different?

File: container/libcontainer/handler.go [hunk 1]
```diff
@@ -499,6 +504,7 @@ func newContainerStats(libcontainerStats *libcontainer.Stats, withPerCPU bool) *
 		setCpuStats(s, ret, withPerCPU)
 		setDiskIoStats(s, ret)
 		setMemoryStats(s, ret)
+		setPidsStats(s, ret)
```
Comment (dashpole) [inline container/libcontainer/handler.go]: I didn't realize this was something we were not already collecting.  Can you add this to ignoremetrics, and ignore it by default?

**Comment by adrien-f** [comment]
I'm sorry for the wait in this PR, I hope to be able to work on it soon :smile:

**Comment by KielChan** [comment]
why not merge before？

**Comment by njuettner** [comment]
@adrien-f @dashpole Any chances we can make some progress here?

**Comment by awprice** [comment]
This would greatly help me out too, looking forward to having these metrics.

**Comment by dashpole** [comment]
Closing this as inactive.  Feel free to reopen if you are still working on this.

**Pull request closed without merging**